in-progress/8025-publishing-tx-objects/design.md

```diff
@@ -48,27 +69,159 @@ To the DA oracle contract, the proposer will publish the following data:
+    proposal_log = self.proposals[bn]
+    if proposal_log.slot <= end_slot:
+      proposal_hashes.append(proposal_log.proposal_hash)
+      block_number = bn
+    else:
+      break
+
+  assert len(proposal_hashes) > 0, 'no proposals'
+
+  prev_archive = self.proposals[self.state.proven_block_count - 1]
+  assert proof.verify(prev_archive, proposal_hashes, archive)
+
+  self.payout_fees();
+
+  self.state.proven_slot = proposal_log[-1].slot
+  self.state.proven_block_count = block_number + 1
+  self.proposals[block_number] = archive
+```
+
+- TODO:
+  - What are we going to do with stuff like the outhashes? They depend on the execution, so we cannot really do it at the time of the proposal if that is. So we would practically need it to happen with the entire epoch at this point. Might just be another tree to make it cheaper. It gets kinda funky with finding the block number because it depends.
 
 ### Forced inclusion of transactions
 
+- Need to fix some notation on what is the "full" proposal etc.
+- My syntax is not really valid vyper but I just enjoy writing it that way, sorry not sorry.
+
+```python
+struct ForceInclusion:
+  nullifier: bytes32
+  include_by_slot: uint256
+  included: bool
+
+struct ForceInclusionProof:
+  proposal: Proposal
+  forced_inclusion_index: uint256, 
+  block_number: uint256
+  membership_proof: bytes32[]
+
+forced_inclusions: public(HashMap[uint256, ForceInclusion])
+forced_inclusion_tip: public(uint256)
+forced_inclusion_count: public(uint256)
+
+FORCE_INCLUSION_DEADLINE: immutable(uint256)
+
+def initiate_force_include(tx, proof, block_number_proven_against):
+  archive = self.proposals[block_number_proven_against].archive
+  assert proof.verify(archive, tx)
+
+  self.forced_inclusions[self.forced_inclusion_count] = ForceInclusion(
+    nullifier = tx.nullifiers[0], 
+    include_by_slot = get_current_slot() + FORCE_INCLUSION_DEADLINE
+  )
+  self.forced_inclusion_count += 1
+
+def show_included(proposal, forced_inclusion_index, block_number, membership_proof):
+  nullifier = self.forced_inclusions[forced_inclusion_index].nullifier
+
+  assert self.proposals[block_number].proposal_hash == proposal.hash()
+  assert membership_proof.verify( nullifier,proposal.txs_hash)
+
+  self.forced_inclusions[forced_inclusion_index].nullifier.included = True
+
+  self.progress_forced_inclusion_tip() 
+
+def progress_forced_inclusion_tip():
+  cache = self.forced_inclusion_tip
+  for i in range(cache, self.forced_inclusion_count):
+    if not self.forced_inclusions[i].included:
+      return
+    self.forced_inclusion_tip = cache
+
+def submit_proof(proof, archive):
+  # super.submit_proof(proof, archive)
+  
+  forced_tip = self.forced_inclusions[self.forced_inclusions_tip]
+  assert forced_tip.included_by_slot == 0 or forced_tip.included_by_slot > proposal_hashes[-1].slot, 'force'
+
+def submit_proof_with_force(proof, archive, force_inclusion_proofs):
```

Comment:
When/where/how does this get called?

Reply:
Will add a comment, but you can see that if a proof is made that includes things that must be forced, you would use it to showcase that it was included, otherwise it might fail due checks at the end

Reply:
So the proposer needs to make the determination of which function to hit in the rollup contract?

Reply:
No, the proof submitter.